### DISCUSSION TITLE
Observing slower response over time for Dropwizard application

### DISCUSSION THREAD

**Discussion by sandeepnkulkarni** [dropwizard/dropwizard#5876] [General]
Our application based on Dropwizard is showing slower response over time after around 1 millions requests are processed. To overcome it, only option we have so far is to restart the application.

I tried to find the root cause of the issue but currently my guess it is something do with either Dropwizard or Jetty. I have tried to use different garbage collectors and does not show much improvements. Though with Parallel GC it seems to be working best. G1 GC and CMS GC are proving to be worse than that. I have analyzed heap dumps and does not show any significant size consumed by remaining objects.

I have created a sample application using 2.1.1 which also shows the similar issue though at much later threshold. It does two things:

1. Generate prime number upto a range 

1. Generate random strings and populate a list with them and then sort and shuffle the list

With above I am able to cause enough CPU and memory consumption which can simulate what our application does in reality.

I have taken 4 runs of 5 million requests each. Runs taken using Corretto 8u332 on 16 Core, 32 GB RAM VM running Windows Server 2019. As shown below, the responses gets slower and slower over time.

**Run 1:** 3 hours 48 minutes
**Run 2:** 4 hours 29 minutes
**Run 3:** 5 hours 13 minutes
**Run 4:** 5 hours 2 minutes

During each run I can see around 95% or above CPU usage, 800-900 MB memory, 100 threads, handle count 1000.

Only 4th run did not show any further degradation in response time. Run 4 was also taken after letting the application rest for a while (few hours). Earlier 3 runs were taken one after the other in succession.

If I disable generation of random strings and only generate prime numbers, things do improve a lot. Each run completes in around 12 mins and there is not much deviations observed in response time. But it does not reflect real life scenario for our application.

I can share the sample application as well in sometime, need to clean up few things before publishing it to my Github account.

Is there any known limitation to Dropwizard and/or Jetty which could be causing this behaviour? Also have a question around higher  handle counts observed, is it common?

**Comment by zUniQueX** [comment]
One resource problem I'm aware of is a bug in Logback, that can cause high CPU overhead with Jetty request logging. This bug is fixed in the Logback 1.2.x branch, but there will probably be no further 1.2.x release.

You can check, if this is the limiting factor, if you switch to the Jetty `CustomRequestLog` implementation.

**Reply by sandeepnkulkarni** [reply]
@zUniQueX I believe you are referring to https://jira.qos.ch/browse/LOGBACK-1580. I will try Jetty `CustomRequestLog` implementation.

**Reply by sandeepnkulkarni** [reply]
We already make use of following configuration for our application:

```
server:
  # by default all requests are written to the console appender which is extremely verbose! Override by specifying an empty requestLog appenders set.
  requestLog:
    appenders: []
```

I have used the same as well in my sample application and test results are with this configuration put in place.

Will it still cause the issue in Logback for Jetty request logging?

**Comment by sandeepnkulkarni** [comment]
Sample application is now put into repository: https://github.com/sandeepk-veritas/dropwizardPerfTest

Client application used to test it does not flood the service, instead makes use of fixed thread pull to submit the request like below:

```
ExecutorService executorService = Executors.newFixedThreadPool(16);
```

Also HTTP client is created with ApacheConnectorProvider.

**Reply by sandeepnkulkarni** [reply]
We again revisited this issue and started doing some more investigation.

We have found that use of Jetty QoSFilter could be a possible reason for slowness observed. We make use of of it because we want to limit the number of parallel processing being done at a given point of time. In our actual application, it is really necessary because memory and CPU usage can grow extensively if number of requests exceeds a certain range.

Sample application shared earlier also uses it (https://github.com/sandeepk-veritas/dropwizardPerfTest/blob/11b1952a75a0e28344db8d608d7f793188863fcd/src/main/java/local/snk/DropwizardPerfTestApplication.java#L120). If we remove it and solely rely on client application to submit requests, at least slowness it not observed to that extent. 

Any thoughts on this observation?
